File: web/teslausb_web/services/storage_config.py
"""Unified storage + cleanup configuration (`/etc/teslausb/teslausb.toml`).

Single source of truth for:

* Partition-reported sizes (TeslaCam, Media) — operator-tunable via the
  web UI. The single `/etc/teslausb/teslafat-0.toml` DiskConfig (one disk,
  two exFAT partitions) is DERIVED from this file by the resize helper
  (`AC.3`), which rewrites the matching `[[partition]]` size.
* The safety buffer (`safety_buffer_gb`) — held back on top of the
  *measured* OS/non-partition SD usage so partition allocation can
  never starve the rootfs. Documented in `docs/06-OPERATIONS.md`;
  operators may edit by hand but the web UI enforces the same minimum.
* The auto-cleanup knobs consumed by the Rust worker
  (`teslausb-worker/src/cleanup.rs`): `target_free_pct`,
  `sentry_max_age_days`, `preserve_with_gps`.

Why a new TOML file instead of folding into `teslausb-web.toml`?
The same data must be read by BOTH the Flask process AND the Rust
worker (which already loads its own `worker.toml`). Putting it in a
dedicated file keeps each consumer's primary config small and lets
the worker reload storage settings on a `SIGHUP` without re-parsing
the entire web config.

Atomic-write contract: callers MUST go through `save()`, which
writes to a sibling `.tmp` file and `os.replace()`s it into place.
Partial writes are therefore invisible to the worker, which polls
this file.
"""

from __future__ import annotations

import logging
import threading
import tomllib
from dataclasses import dataclass, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final

if TYPE_CHECKING:
    from collections.abc import Mapping

logger = logging.getLogger(__name__)

DEFAULT_CONFIG_PATH: Final[Path] = Path("/etc/teslausb/teslausb.toml")

# Safety buffer floor — held back ON TOP OF the *measured* OS/non-partition
# SD usage so the card can never physically reach 0 (which would crash the
# rootfs and stop TeslaCam writes). This is the OS-growth budget cushion:
# even after both partitions fill to their advertised caps, at least this
# many GB stay free for journal/index/swap growth. Replaces the legacy flat
# ``os_reserve_gb`` (which assumed a fixed OS footprint); the reserve is now
# ``measured_os_usage + safety_buffer_gb``. Matches ``SAFETY_BUFFER_*`` in
# ``rust/.../storage_config.rs`` and the ``teslausb-resize-lun`` helper.
SAFETY_BUFFER_MIN_GB: Final[int] = 5
SAFETY_BUFFER_DEFAULT_GB: Final[int] = 5

# LUN size bounds match the teslafat backend's accepted range
# (`setup-lib/11-gadget.sh::_b1_validate_size_gb`).
LUN_MIN_GB: Final[int] = 4
LUN_MAX_GB: Final[int] = 2048

# Free-space target bounds. 0 = auto-tune from indexer median.
TARGET_FREE_PCT_MIN: Final[int] = 0
TARGET_FREE_PCT_MAX: Final[int] = 50

# Sentry max-age bounds. 0 = unlimited (never auto-delete based on
# age alone; sentry is still last-resort fodder when A+B exhausted).
SENTRY_MAX_AGE_MIN: Final[int] = 0
SENTRY_MAX_AGE_MAX: Final[int] = 3650


class StorageConfigError(ValueError):
    """Storage configuration is invalid or could not be parsed."""


@dataclass(frozen=True, slots=True)
class StorageSection:
    """Partition sizing + safety-buffer guard.

    ``safety_buffer_gb`` is held back on top of the *measured* OS/non-
    partition SD usage; the resize cap enforces
    ``teslacam_gb + media_gb + measured_os_usage + safety_buffer_gb
    <= sd_total_gb`` so the card can never physically overfill.
    """

    safety_buffer_gb: int = SAFETY_BUFFER_DEFAULT_GB
    teslacam_gb: int = 64
    media_gb: int = 32


@dataclass(frozen=True, slots=True)
class CleanupSection:
    """Auto-cleanup knobs consumed by the Rust worker."""

    target_free_pct: int = 0
    sentry_max_age_days: int = 0
    preserve_with_gps: bool = True


@dataclass(frozen=True, slots=True)
class TeslausbConfig:
    """Top-level snapshot of `/etc/teslausb/teslausb.toml`."""

    storage: StorageSection
    cleanup: CleanupSection


_LOCK = threading.RLock()


def default_config() -> TeslausbConfig:
    """Return the built-in defaults (used when the file is absent)."""
    return TeslausbConfig(storage=StorageSection(), cleanup=CleanupSection())


def load(path: Path | None = None) -> TeslausbConfig:
    """Read and validate the config. Returns defaults if file is absent.

    Raises `StorageConfigError` on parse failure or invalid values.
    Bounds-checks every field but does NOT enforce the cross-field
    `teslacam + media + os_usage + safety_buffer <= sd_total` constraint
    — that requires knowing the SD card capacity and is checked by
    `validate_against_capacity()`.
    """
    target = path or DEFAULT_CONFIG_PATH
    if not target.exists():
        return default_config()

    try:
        with target.open("rb") as handle:
            payload = tomllib.load(handle)
    except (OSError, tomllib.TOMLDecodeError) as exc:
        raise StorageConfigError(f"failed to read {target}: {exc}") from exc

    storage = _parse_storage(payload.get("storage", {}))
    cleanup = _parse_cleanup(payload.get("cleanup", {}))
    return TeslausbConfig(storage=storage, cleanup=cleanup)


def save(config: TeslausbConfig, path: Path | None = None) -> None:
    """Atomically write `config` to `path`. Creates parent dir if missing.

    Performs the same bounds checks as `load()` so a programmatic
    caller can't write something `load()` would later reject.
    """
    _validate_storage(config.storage)
    _validate_cleanup(config.cleanup)
    target = path or DEFAULT_CONFIG_PATH
    body = _render(config)
    with _LOCK:
        target.parent.mkdir(parents=True, exist_ok=True)
        tmp = target.with_suffix(target.suffix + ".tmp")
        tmp.write_text(body, encoding="utf-8")
        # Mode 0664 so the file remains group-writable for the
        # `teslausb` group after a subsequent atomic rename. The
        # web service runs as a member of that group (see setup-lib
        # `_install_storage_config_perms`); without this chmod the
        # default umask would yield 0644 and the next save() would
        # fail with EACCES.
        tmp.chmod(0o664)
        tmp.replace(target)
    logger.info("storage_config: wrote %s", target)


def validate_against_capacity(
    config: TeslausbConfig,
    sd_total_gb: int,
    os_usage_gb: int = 0,
) -> None:
    """Cross-field check: partitions + measured OS usage + safety buffer
    must fit within the SD card.

    The reserve held back is ``os_usage_gb`` (the *measured* non-partition
    SD usage: OS install, journal, swap, index, tmp) plus the configured
    ``safety_buffer_gb`` cushion. Enforcing
    ``teslacam + media + os_usage + safety_buffer <= sd_total`` guarantees
    that even if both partitions fill to their advertised caps, at least
    ``safety_buffer_gb`` stays physically free, so the card can never crash
    the rootfs.

    Raises `StorageConfigError` with a human-readable message on failure.
    `sd_total_gb=0` is treated as "unknown" and skips the check (e.g., in
    unit tests with no backing filesystem). `os_usage_gb` defaults to 0 so
    callers without a measurement still get the buffer-only floor.
    """
    if sd_total_gb <= 0:
        return
    storage = config.storage
    reserve = max(0, os_usage_gb) + storage.safety_buffer_gb
    usable = sd_total_gb - reserve
    requested = storage.teslacam_gb + storage.media_gb
    if requested > usable:
        raise StorageConfigError(
            f"teslacam_gb + media_gb = {requested} GB exceeds usable "
            f"capacity {usable} GB (sd_total={sd_total_gb} GB minus "
            f"measured OS usage {max(0, os_usage_gb)} GB minus safety "
            f"buffer {storage.safety_buffer_gb} GB)",
        )


def with_storage(
    config: TeslausbConfig,
    *,
    safety_buffer_gb: int | None = None,
    teslacam_gb: int | None = None,
    media_gb: int | None = None,
) -> TeslausbConfig:
    """Return a copy with the named storage fields replaced."""
    updated = replace(
        config.storage,
        safety_buffer_gb=(
            safety_buffer_gb if safety_buffer_gb is not None else config.storage.safety_buffer_gb
        ),
        teslacam_gb=teslacam_gb if teslacam_gb is not None else config.storage.teslacam_gb,
        media_gb=media_gb if media_gb is not None else config.storage.media_gb,
    )
    return replace(config, storage=updated)


def with_cleanup(
    config: TeslausbConfig,
    *,
    target_free_pct: int | None = None,
    sentry_max_age_days: int | None = None,
    preserve_with_gps: bool | None = None,
) -> TeslausbConfig:
    """Return a copy with the named cleanup fields replaced."""
    updated = replace(
        config.cleanup,
        target_free_pct=(
            target_free_pct if target_free_pct is not None else config.cleanup.target_free_pct
        ),
        sentry_max_age_days=(
            sentry_max_age_days
            if sentry_max_age_days is not None
            else config.cleanup.sentry_max_age_days
        ),
        preserve_with_gps=(
            preserve_with_gps if preserve_with_gps is not None else config.cleanup.preserve_with_gps
        ),
    )
    return replace(config, cleanup=updated)


# --------------------------------------------------------------------------- #
# internals
# --------------------------------------------------------------------------- #


def _parse_storage(section: Mapping[str, object]) -> StorageSection:
    # Back-compat: a pre-existing teslausb.toml (before the safety-buffer
    # rework) carries ``os_reserve_gb`` instead of ``safety_buffer_gb``.
    # Map the legacy value onto the buffer (a larger held-back value is
    # always conservative/safe) so an old file still loads; the next save()
    # rewrites it with the new key.
    if "safety_buffer_gb" in section:
        buffer_gb = _as_int(section, "safety_buffer_gb", SAFETY_BUFFER_DEFAULT_GB)
    elif "os_reserve_gb" in section:
        buffer_gb = _as_int(section, "os_reserve_gb", SAFETY_BUFFER_DEFAULT_GB)
    else:
        buffer_gb = SAFETY_BUFFER_DEFAULT_GB
    candidate = StorageSection(
        safety_buffer_gb=buffer_gb,
        teslacam_gb=_as_int(section, "teslacam_gb", StorageSection().teslacam_gb),
        media_gb=_as_int(section, "media_gb", StorageSection().media_gb),
    )
    _validate_storage(candidate)
    return candidate


def _parse_cleanup(section: Mapping[str, object]) -> CleanupSection:
    candidate = CleanupSection(
        target_free_pct=_as_int(section, "target_free_pct", 0),
        sentry_max_age_days=_as_int(section, "sentry_max_age_days", 0),
        preserve_with_gps=_as_bool(section, "preserve_with_gps", default=True),
    )
    _validate_cleanup(candidate)
    return candidate


def _validate_storage(section: StorageSection) -> None:
    if section.safety_buffer_gb < SAFETY_BUFFER_MIN_GB:
        raise StorageConfigError(
            f"safety_buffer_gb must be >= {SAFETY_BUFFER_MIN_GB}, "
            f"got {section.safety_buffer_gb}",
        )
    _check_lun("teslacam_gb", section.teslacam_gb)
    _check_lun("media_gb", section.media_gb)


def _validate_cleanup(section: CleanupSection) -> None:
    if not TARGET_FREE_PCT_MIN <= section.target_free_pct <= TARGET_FREE_PCT_MAX:
        raise StorageConfigError(
            f"target_free_pct must be in "
            f"[{TARGET_FREE_PCT_MIN}, {TARGET_FREE_PCT_MAX}], "
            f"got {section.target_free_pct}",
        )
    if not SENTRY_MAX_AGE_MIN <= section.sentry_max_age_days <= SENTRY_MAX_AGE_MAX:
        raise StorageConfigError(
            f"sentry_max_age_days must be in "
            f"[{SENTRY_MAX_AGE_MIN}, {SENTRY_MAX_AGE_MAX}], "
            f"got {section.sentry_max_age_days}",
        )


def _check_lun(field_name: str, value: int) -> None:
    if not LUN_MIN_GB <= value <= LUN_MAX_GB:
        raise StorageConfigError(
            f"{field_name} must be in [{LUN_MIN_GB}, {LUN_MAX_GB}], got {value}",
        )


def _as_int(section: Mapping[str, object], key: str, default: int) -> int:
    if key not in section:
        return default
    value = section[key]
    if isinstance(value, bool) or not isinstance(value, int):
        raise StorageConfigError(f"{key} must be an integer, got {value!r}")
    return value


def _as_bool(section: Mapping[str, object], key: str, *, default: bool) -> bool:
    if key not in section:
        return default
    value = section[key]
    if not isinstance(value, bool):
        raise StorageConfigError(f"{key} must be a boolean, got {value!r}")
    return value


def _render(config: TeslausbConfig) -> str:
    storage = config.storage
    cleanup = config.cleanup
    return (
        "# Managed by teslausb-b1 (web UI / setup.sh).\n"
        "# Documented in docs/06-OPERATIONS.md.\n"
        "\n"
        "[storage]\n"
        f"safety_buffer_gb = {storage.safety_buffer_gb}\n"
        f"teslacam_gb = {storage.teslacam_gb}\n"
        f"media_gb = {storage.media_gb}\n"
        "\n"
        "[cleanup]\n"
        f"target_free_pct = {cleanup.target_free_pct}\n"
        f"sentry_max_age_days = {cleanup.sentry_max_age_days}\n"
        f"preserve_with_gps = {'true' if cleanup.preserve_with_gps else 'false'}\n"
    )

File: rust/crates/teslausb-worker/src/storage_config.rs
//! Loader for the unified `/etc/teslausb/teslausb.toml` config
//! introduced in AC.1.
//!
//! Mirrors the Python schema in
//! `web/teslausb_web/services/storage_config.py`. Owned here in
//! the Rust worker because the cleanup loop (AC.5) consumes the
//! `[cleanup]` section on every sweep and the resize helper
//! (AC.3) consumes the `[storage]` section on every apply.
//!
//! Why a second config file (not a section in `worker.toml`)?
//! The same bytes must be read by BOTH the worker AND the Flask
//! UI, and the UI writes back through `storage_config.py`'s
//! `save()`. Sharing one file keeps the two consumers' views in
//! sync without an IPC round-trip on every page load.
//!
//! Backwards-compat: callers fall back to
//! [`StorageConfig::default()`] when the file is absent. The
//! legacy `/etc/teslausb/teslafat-{0,1}.toml` files remain the
//! source of truth for `teslafat` (different schema, different
//! consumer). The AC.3 resize helper regenerates teslafat-*.toml
//! from teslausb.toml so they never diverge in practice.

// "TeslaCam", "exFAT", "FAT32", "SEI", "GPS", "TOML" are domain
// terms; backticking every mention adds noise without value.
// Matches the SEI / cleanup carve-out.
#![allow(clippy::doc_markdown)]

use std::path::Path;

use anyhow::{Context, Result, ensure};
use serde::Deserialize;

/// Lower bound for `safety_buffer_gb` — held back on top of the
/// *measured* OS/non-partition SD usage so partition allocation can
/// never starve the rootfs. Matches `SAFETY_BUFFER_MIN_GB` in
/// `storage_config.py`.
pub const SAFETY_BUFFER_MIN_GB: u32 = 5;
/// Default for `safety_buffer_gb`. Matches `SAFETY_BUFFER_DEFAULT_GB`
/// in `storage_config.py`.
pub const SAFETY_BUFFER_DEFAULT_GB: u32 = 5;
/// Lower bound for any per-LUN size, in GB.
pub const LUN_MIN_GB: u32 = 4;
/// Upper bound for any per-LUN size, in GB. Matches the
/// teslafat backend's accepted range.
pub const LUN_MAX_GB: u32 = 2048;
/// Upper bound for `target_free_pct`. `0` is the auto-tune
/// sentinel (cleanup loop computes 2× bytes-per-recording-minute
/// from the indexer median).
pub const TARGET_FREE_PCT_MAX: u8 = 50;
/// Upper bound for `sentry_max_age_days`. `0` = unlimited
/// (Sentry only auto-deleted as a last resort).
pub const SENTRY_MAX_AGE_DAYS_MAX: u32 = 3650;

const fn default_safety_buffer_gb() -> u32 {
    SAFETY_BUFFER_DEFAULT_GB
}

const fn default_teslacam_gb() -> u32 {
    64
}

const fn default_media_gb() -> u32 {
    32
}

const fn default_zero_u8() -> u8 {
    0
}

const fn default_zero_u32() -> u32 {
    0
}

const fn default_true() -> bool {
    true
}

/// Partition sizing + safety-buffer guard. Mirrors the `[storage]`
/// section of teslausb.toml.
#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct StorageSection {
    /// Cushion held back on top of the *measured* OS/non-partition SD
    /// usage so partition allocation can never starve the rootfs.
    /// Accepts the legacy `os_reserve_gb` key for back-compat with a
    /// pre-rework teslausb.toml (the web layer rewrites it on next save).
    #[serde(default = "default_safety_buffer_gb", alias = "os_reserve_gb")]
    pub safety_buffer_gb: u32,
    /// Size partition 0 reports to Tesla (TeslaCam, exFAT).
    #[serde(default = "default_teslacam_gb")]
    pub teslacam_gb: u32,
    /// Size partition 1 reports to Tesla (media, exFAT).
    #[serde(default = "default_media_gb")]
    pub media_gb: u32,
}

impl Default for StorageSection {
    fn default() -> Self {
        Self {
            safety_buffer_gb: SAFETY_BUFFER_DEFAULT_GB,
            teslacam_gb: default_teslacam_gb(),
            media_gb: default_media_gb(),
        }
    }
}

/// Auto-cleanup knobs consumed by `cleanup.rs` (AC.5). Mirrors
/// the `[cleanup]` section of teslausb.toml.
#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct CleanupSection {
    /// Percent of TeslaCam capacity to keep free. `0` means
    /// auto-tune: 2× the median 6-camera-1-minute recording
    /// size, expressed as a percent of LUN capacity.
    #[serde(default = "default_zero_u8")]
    pub target_free_pct: u8,
    /// Sentry events older than this become eligible for
    /// auto-deletion via Tier C. `0` = unlimited (Sentry only
    /// deleted when Tier A+B exhausted and free still below
    /// target).
    #[serde(default = "default_zero_u32")]
    pub sentry_max_age_days: u32,
    /// When true, `RecentClips` with GPS or SEI tesla-data are
    /// classified Tier B (preserved over plain clips).
    #[serde(default = "default_true")]
    pub preserve_with_gps: bool,
}

impl Default for CleanupSection {
    fn default() -> Self {
        Self {
            target_free_pct: 0,
            sentry_max_age_days: 0,
            preserve_with_gps: true,
        }
    }
}

/// Full snapshot of `/etc/teslausb/teslausb.toml`.
#[derive(Debug, Clone, Default, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct StorageConfig {
    /// LUN sizing section.
    #[serde(default)]
    pub storage: StorageSection,
    /// Cleanup section.
    #[serde(default)]
    pub cleanup: CleanupSection,
}

impl StorageConfig {
    /// Load and validate a config from `path`. Returns
    /// [`StorageConfig::default()`] if the file is absent
    /// (back-compat: a freshly-installed device without AC.1's
    /// setup step should still boot).
    ///
    /// # Errors
    ///
    /// Returns `Err` if the file exists but cannot be read,
    /// the bytes are not valid TOML, an unknown field is
    /// present, or semantic validation rejects a value.
    pub fn load(path: &Path) -> Result<Self> {
        if !path.exists() {
            return Ok(Self::default());
        }
        let raw =
            std::fs::read_to_string(path).with_context(|| format!("reading {}", path.display()))?;
        let cfg: Self =
            toml::from_str(&raw).with_context(|| format!("parsing {}", path.display()))?;
        cfg.validate()?;
        Ok(cfg)
    }

    /// Semantic validation. Bounds-checks every field but does
    /// NOT enforce the cross-field `teslacam + media + os_usage +
    /// safety_buffer <= sd_total` constraint.
    ///
    /// The worker only *reads* this config (for cleanup tuning); it never
    /// resizes partitions, so it is deliberately NOT one of the layers that
    /// enforce the no-overcommit capacity invariant. The two enforcing
    /// layers are the web `apply_storage_config` pre-check and the
    /// authoritative `teslausb-resize-lun` bash helper (which re-samples
    /// `df`/`du` at apply time). Here we only guarantee the values are
    /// individually in range so the worker never acts on a garbage size.
    ///
    /// # Errors
    ///
    /// Returns `Err` with a `storage.<field>` or `cleanup.<field>`
    /// prefix identifying the offending value.
    pub fn validate(&self) -> Result<()> {
        ensure!(
            self.storage.safety_buffer_gb >= SAFETY_BUFFER_MIN_GB,
            "storage.safety_buffer_gb must be >= {SAFETY_BUFFER_MIN_GB} (got {})",
            self.storage.safety_buffer_gb,
        );
        check_lun_size("storage.teslacam_gb", self.storage.teslacam_gb)?;
        check_lun_size("storage.media_gb", self.storage.media_gb)?;
        ensure!(
            self.cleanup.target_free_pct <= TARGET_FREE_PCT_MAX,
            "cleanup.target_free_pct must be in [0, {TARGET_FREE_PCT_MAX}] (got {})",
            self.cleanup.target_free_pct,
        );
        ensure!(
            self.cleanup.sentry_max_age_days <= SENTRY_MAX_AGE_DAYS_MAX,
            "cleanup.sentry_max_age_days must be in [0, {SENTRY_MAX_AGE_DAYS_MAX}] (got {})",
            self.cleanup.sentry_max_age_days,
        );
        Ok(())
    }
}

fn check_lun_size(name: &str, value: u32) -> Result<()> {
    ensure!(
        (LUN_MIN_GB..=LUN_MAX_GB).contains(&value),
        "{name} must be in [{LUN_MIN_GB}, {LUN_MAX_GB}] (got {value})",
    );
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn defaults_round_trip() {
        let cfg: StorageConfig = toml::from_str("").unwrap();
        cfg.validate().unwrap();
        assert_eq!(cfg, StorageConfig::default());
    }

    #[test]
    fn parses_full_file() {
        let raw = "\
[storage]
safety_buffer_gb = 12
teslacam_gb = 128
media_gb = 64

[cleanup]
target_free_pct = 12
sentry_max_age_days = 90
preserve_with_gps = false
";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        cfg.validate().unwrap();
        assert_eq!(cfg.storage.safety_buffer_gb, 12);
        assert_eq!(cfg.storage.teslacam_gb, 128);
        assert_eq!(cfg.storage.media_gb, 64);
        assert_eq!(cfg.cleanup.target_free_pct, 12);
        assert_eq!(cfg.cleanup.sentry_max_age_days, 90);
        assert!(!cfg.cleanup.preserve_with_gps);
    }

    #[test]
    fn accepts_legacy_os_reserve_alias() {
        // A pre-rework teslausb.toml carries os_reserve_gb; it must still
        // parse (mapped onto safety_buffer_gb) until the web layer rewrites
        // it with the new key on the next save.
        let raw = "[storage]\nos_reserve_gb = 20\n";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        cfg.validate().unwrap();
        assert_eq!(cfg.storage.safety_buffer_gb, 20);
    }

    #[test]
    fn load_returns_defaults_when_file_absent() {
        let tmp = tempfile::tempdir().unwrap();
        let cfg = StorageConfig::load(&tmp.path().join("missing.toml")).unwrap();
        assert_eq!(cfg, StorageConfig::default());
    }

    #[test]
    fn load_parses_file_when_present() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("teslausb.toml");
        std::fs::write(
            &path,
            "[storage]\nteslacam_gb = 100\n[cleanup]\nsentry_max_age_days = 60\n",
        )
        .unwrap();
        let cfg = StorageConfig::load(&path).unwrap();
        assert_eq!(cfg.storage.teslacam_gb, 100);
        assert_eq!(cfg.cleanup.sentry_max_age_days, 60);
    }

    #[test]
    fn rejects_low_safety_buffer() {
        let raw = "[storage]\nsafety_buffer_gb = 4\n";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        let err = cfg.validate().unwrap_err();
        assert!(err.to_string().contains("safety_buffer_gb"), "got: {err}",);
    }

    #[test]
    fn rejects_lun_below_min() {
        let raw = "[storage]\nteslacam_gb = 1\n";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        let err = cfg.validate().unwrap_err();
        assert!(err.to_string().contains("teslacam_gb"), "got: {err}");
    }

    #[test]
    fn rejects_lun_above_max() {
        let raw = "[storage]\nmedia_gb = 9999\n";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        let err = cfg.validate().unwrap_err();
        assert!(err.to_string().contains("media_gb"), "got: {err}");
    }

    #[test]
    fn rejects_target_free_pct_above_max() {
        let raw = "[cleanup]\ntarget_free_pct = 80\n";
        let cfg: StorageConfig = toml::from_str(raw).unwrap();
        let err = cfg.validate().unwrap_err();
        assert!(err.to_string().contains("target_free_pct"), "got: {err}");
    }

    #[test]
    fn rejects_unknown_storage_field() {
        let raw = "[storage]\nbogus = 1\n";
        let err = toml::from_str::<StorageConfig>(raw).unwrap_err();
        assert!(err.to_string().contains("bogus"), "got: {err}");
    }

    #[test]
    fn rejects_unknown_cleanup_field() {
        let raw = "[cleanup]\nbogus = 1\n";
        let err = toml::from_str::<StorageConfig>(raw).unwrap_err();
        assert!(err.to_string().contains("bogus"), "got: {err}");
    }
}
